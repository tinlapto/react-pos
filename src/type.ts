export type Item = {
  id: number;
  title: string;
  photo: string;
  price: number;
};

export type Topping = {
  id: number;
  name: string;
};

export type Size = {
  id: number;
  name: string;
};

export type ItemDetail = Item & {
  toppings: Topping[];
  sizes: Size[];
};

export type CartItem = {
  id: number;
  title: string;
  price: number;
  toppings: Topping[];
  sizes: Size[];
  selectedToppings: number[];
  selectedSize: number;
  count: number;
};
