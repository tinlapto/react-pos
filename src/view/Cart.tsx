import { Button, Col, Divider, Popconfirm, Row } from "antd";
import { CloseOutlined, ShoppingCartOutlined } from "@ant-design/icons";
import CartItem from "../component/CartItem";
import { CartItem as TypeCartItem } from "../type";

const Cart = ({
  items,
  setItems,
}: {
  items: TypeCartItem[];
  setItems: (items: TypeCartItem[]) => void;
}) => {
  return (
    <div style={{ display: "flex", flexDirection: "column", height: "100%" }}>
      <div style={{ flex: 1, overflow: "auto" }}>
        {items.length === 0 && (
          <Row>
            <Col span={24} style={{ textAlign: "center", padding: 20 }}>
              No items
            </Col>
          </Row>
        )}
        {items.map((item, index) => (
          <CartItem
            key={index}
            item={item}
            onRemove={() => setItems(items.filter((_, i) => i !== index))}
          />
        ))}
        <Divider />
        <Row justify="end" gutter={10}>
          <Col style={{ textAlign: "right" }} span={10}>
            Total:
          </Col>
          <Col span={4}>
            ${items.reduce((v, item) => item.price * item.count + v, 0)}
          </Col>
        </Row>
      </div>
      <div style={{ display: "flex" }}>
        <Button
          style={{ flex: 1, marginRight: 5 }}
          type="primary"
          size={"large"}
          icon={<ShoppingCartOutlined />}
          disabled={items.length === 0}
        >
          Checkout
        </Button>
        <Popconfirm
          title={"Confirm to remove all items?"}
          onConfirm={() => setItems([])}
          okText={"Remove"}
          okButtonProps={{ danger: true }}
          disabled={items.length === 0}
        >
          <Button
            type="primary"
            size={"large"}
            disabled={items.length === 0}
            danger
            icon={<CloseOutlined />}
          />
        </Popconfirm>
      </div>
    </div>
  );
};

export default Cart;
