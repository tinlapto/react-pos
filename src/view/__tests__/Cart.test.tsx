import { render, screen, fireEvent } from "@testing-library/react";
import Cart from "../Cart";
import { CartItem as TypeCartItem } from "../../type";

const items: TypeCartItem[] = [
  {
    id: 1,
    title: "test item",
    toppings: [{ id: 1, name: "topping 1" }],
    sizes: [{ id: 1, name: "size 1" }],
    selectedToppings: [1],
    selectedSize: 1,
    count: 1,
    price: 10,
  },
  {
    id: 2,
    title: "test item",
    toppings: [{ id: 1, name: "topping 1" }],
    sizes: [{ id: 1, name: "size 1" }],
    selectedToppings: [1],
    selectedSize: 1,
    count: 1,
    price: 10,
  },
];

describe("Cart", () => {
  beforeAll(() => {
    // @ts-ignore
    delete window.matchMedia;
    window.matchMedia = (query) => ({
      matches: false,
      media: query,
      onchange: null,
      addListener: jest.fn(), // deprecated
      removeListener: jest.fn(), // deprecated
      addEventListener: jest.fn(),
      removeEventListener: jest.fn(),
      dispatchEvent: jest.fn(),
    });
  });
  test("render remove one", () => {
    const onSetItems = jest.fn();
    const { asFragment } = render(
      <Cart items={items} setItems={onSetItems} />,
      { container: document.body }
    );
    expect(asFragment()).toMatchSnapshot();
    fireEvent.click(screen.getAllByRole("img")[0]);
    fireEvent.click(screen.getByText("Remove"));
    expect(onSetItems).toBeCalledWith(items.slice(1));
  });

  test("render remove all", () => {
    const onSetItems = jest.fn();
    const { asFragment } = render(
      <Cart items={items} setItems={onSetItems} />,
      { container: document.body }
    );
    expect(asFragment()).toMatchSnapshot();
    const imgs = screen.getAllByRole("img");
    fireEvent.click(imgs[imgs.length - 1]);
    fireEvent.click(screen.getByText("Remove"));
    expect(onSetItems).toBeCalledWith([]);
  });
});
