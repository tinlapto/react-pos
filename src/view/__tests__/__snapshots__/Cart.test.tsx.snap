// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart render remove all 1`] = `
<DocumentFragment>
  <div
    style="display: flex; flex-direction: column; height: 100%;"
  >
    <div
      style="flex: 1; overflow: auto;"
    >
      <div
        class="ant-row"
        style="margin: 5px 0px; border-bottom: 0.1px solid #eee; padding: 5px 0px;"
      >
        <div
          class="ant-col ant-col-4"
          style="padding-left: 5px; padding-right: 5px; display: flex; align-items: center;"
        >
          <button
            class="ant-btn ant-btn-default ant-btn-sm ant-btn-icon-only ant-btn-dangerous"
            type="button"
          >
            <span
              aria-label="close"
              class="anticon anticon-close"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="close"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M563.8 512l262.5-312.9c4.4-5.2.7-13.1-6.1-13.1h-79.8c-4.7 0-9.2 2.1-12.3 5.7L511.6 449.8 295.1 191.7c-3-3.6-7.5-5.7-12.3-5.7H203c-6.8 0-10.5 7.9-6.1 13.1L459.4 512 196.9 824.9A7.95 7.95 0 00203 838h79.8c4.7 0 9.2-2.1 12.3-5.7l216.5-258.1 216.5 258.1c3 3.6 7.5 5.7 12.3 5.7h79.8c6.8 0 10.5-7.9 6.1-13.1L563.8 512z"
                />
              </svg>
            </span>
          </button>
        </div>
        <div
          class="ant-col ant-col-16"
          style="padding-left: 5px; padding-right: 5px; text-align: right;"
        >
          <div>
            1 X test item
          </div>
          <div>
            <div
              style="color: rgb(170, 170, 170); font-size: 12px;"
            >
              topping 1
            </div>
          </div>
          <div>
            <div
              style="color: rgb(170, 170, 170); font-size: 12px;"
            >
              size 1
            </div>
          </div>
        </div>
        <div
          class="ant-col ant-col-4"
          style="padding-left: 5px; padding-right: 5px; display: flex; align-items: center;"
        >
          $10
        </div>
      </div>
      <div
        class="ant-row"
        style="margin: 5px 0px; border-bottom: 0.1px solid #eee; padding: 5px 0px;"
      >
        <div
          class="ant-col ant-col-4"
          style="padding-left: 5px; padding-right: 5px; display: flex; align-items: center;"
        >
          <button
            class="ant-btn ant-btn-default ant-btn-sm ant-btn-icon-only ant-btn-dangerous"
            type="button"
          >
            <span
              aria-label="close"
              class="anticon anticon-close"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="close"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M563.8 512l262.5-312.9c4.4-5.2.7-13.1-6.1-13.1h-79.8c-4.7 0-9.2 2.1-12.3 5.7L511.6 449.8 295.1 191.7c-3-3.6-7.5-5.7-12.3-5.7H203c-6.8 0-10.5 7.9-6.1 13.1L459.4 512 196.9 824.9A7.95 7.95 0 00203 838h79.8c4.7 0 9.2-2.1 12.3-5.7l216.5-258.1 216.5 258.1c3 3.6 7.5 5.7 12.3 5.7h79.8c6.8 0 10.5-7.9 6.1-13.1L563.8 512z"
                />
              </svg>
            </span>
          </button>
        </div>
        <div
          class="ant-col ant-col-16"
          style="padding-left: 5px; padding-right: 5px; text-align: right;"
        >
          <div>
            1 X test item
          </div>
          <div>
            <div
              style="color: rgb(170, 170, 170); font-size: 12px;"
            >
              topping 1
            </div>
          </div>
          <div>
            <div
              style="color: rgb(170, 170, 170); font-size: 12px;"
            >
              size 1
            </div>
          </div>
        </div>
        <div
          class="ant-col ant-col-4"
          style="padding-left: 5px; padding-right: 5px; display: flex; align-items: center;"
        >
          $10
        </div>
      </div>
      <div
        class="ant-divider ant-divider-horizontal"
        role="separator"
      />
      <div
        class="ant-row ant-row-end"
        style="margin-left: -5px; margin-right: -5px;"
      >
        <div
          class="ant-col ant-col-10"
          style="padding-left: 5px; padding-right: 5px; text-align: right;"
        >
          Total:
        </div>
        <div
          class="ant-col ant-col-4"
          style="padding-left: 5px; padding-right: 5px;"
        >
          $20
        </div>
      </div>
    </div>
    <div
      style="display: flex;"
    >
      <button
        class="ant-btn ant-btn-primary ant-btn-lg"
        style="flex: 1; margin-right: 5px;"
        type="button"
      >
        <span
          aria-label="shopping-cart"
          class="anticon anticon-shopping-cart"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="shopping-cart"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="0 0 1024 1024"
            width="1em"
          >
            <path
              d="M922.9 701.9H327.4l29.9-60.9 496.8-.9c16.8 0 31.2-12 34.2-28.6l68.8-385.1c1.8-10.1-.9-20.5-7.5-28.4a34.99 34.99 0 00-26.6-12.5l-632-2.1-5.4-25.4c-3.4-16.2-18-28-34.6-28H96.5a35.3 35.3 0 100 70.6h125.9L246 312.8l58.1 281.3-74.8 122.1a34.96 34.96 0 00-3 36.8c6 11.9 18.1 19.4 31.5 19.4h62.8a102.43 102.43 0 00-20.6 61.7c0 56.6 46 102.6 102.6 102.6s102.6-46 102.6-102.6c0-22.3-7.4-44-20.6-61.7h161.1a102.43 102.43 0 00-20.6 61.7c0 56.6 46 102.6 102.6 102.6s102.6-46 102.6-102.6c0-22.3-7.4-44-20.6-61.7H923c19.4 0 35.3-15.8 35.3-35.3a35.42 35.42 0 00-35.4-35.2zM305.7 253l575.8 1.9-56.4 315.8-452.3.8L305.7 253zm96.9 612.7c-17.4 0-31.6-14.2-31.6-31.6 0-17.4 14.2-31.6 31.6-31.6s31.6 14.2 31.6 31.6a31.6 31.6 0 01-31.6 31.6zm325.1 0c-17.4 0-31.6-14.2-31.6-31.6 0-17.4 14.2-31.6 31.6-31.6s31.6 14.2 31.6 31.6a31.6 31.6 0 01-31.6 31.6z"
            />
          </svg>
        </span>
        <span>
          Checkout
        </span>
      </button>
      <button
        class="ant-btn ant-btn-primary ant-btn-lg ant-btn-icon-only ant-btn-dangerous"
        type="button"
      >
        <span
          aria-label="close"
          class="anticon anticon-close"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="close"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M563.8 512l262.5-312.9c4.4-5.2.7-13.1-6.1-13.1h-79.8c-4.7 0-9.2 2.1-12.3 5.7L511.6 449.8 295.1 191.7c-3-3.6-7.5-5.7-12.3-5.7H203c-6.8 0-10.5 7.9-6.1 13.1L459.4 512 196.9 824.9A7.95 7.95 0 00203 838h79.8c4.7 0 9.2-2.1 12.3-5.7l216.5-258.1 216.5 258.1c3 3.6 7.5 5.7 12.3 5.7h79.8c6.8 0 10.5-7.9 6.1-13.1L563.8 512z"
            />
          </svg>
        </span>
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Cart render remove one 1`] = `
<DocumentFragment>
  <div
    style="display: flex; flex-direction: column; height: 100%;"
  >
    <div
      style="flex: 1; overflow: auto;"
    >
      <div
        class="ant-row"
        style="margin: 5px 0px; border-bottom: 0.1px solid #eee; padding: 5px 0px;"
      >
        <div
          class="ant-col ant-col-4"
          style="padding-left: 5px; padding-right: 5px; display: flex; align-items: center;"
        >
          <button
            class="ant-btn ant-btn-default ant-btn-sm ant-btn-icon-only ant-btn-dangerous"
            type="button"
          >
            <span
              aria-label="close"
              class="anticon anticon-close"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="close"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M563.8 512l262.5-312.9c4.4-5.2.7-13.1-6.1-13.1h-79.8c-4.7 0-9.2 2.1-12.3 5.7L511.6 449.8 295.1 191.7c-3-3.6-7.5-5.7-12.3-5.7H203c-6.8 0-10.5 7.9-6.1 13.1L459.4 512 196.9 824.9A7.95 7.95 0 00203 838h79.8c4.7 0 9.2-2.1 12.3-5.7l216.5-258.1 216.5 258.1c3 3.6 7.5 5.7 12.3 5.7h79.8c6.8 0 10.5-7.9 6.1-13.1L563.8 512z"
                />
              </svg>
            </span>
          </button>
        </div>
        <div
          class="ant-col ant-col-16"
          style="padding-left: 5px; padding-right: 5px; text-align: right;"
        >
          <div>
            1 X test item
          </div>
          <div>
            <div
              style="color: rgb(170, 170, 170); font-size: 12px;"
            >
              topping 1
            </div>
          </div>
          <div>
            <div
              style="color: rgb(170, 170, 170); font-size: 12px;"
            >
              size 1
            </div>
          </div>
        </div>
        <div
          class="ant-col ant-col-4"
          style="padding-left: 5px; padding-right: 5px; display: flex; align-items: center;"
        >
          $10
        </div>
      </div>
      <div
        class="ant-row"
        style="margin: 5px 0px; border-bottom: 0.1px solid #eee; padding: 5px 0px;"
      >
        <div
          class="ant-col ant-col-4"
          style="padding-left: 5px; padding-right: 5px; display: flex; align-items: center;"
        >
          <button
            class="ant-btn ant-btn-default ant-btn-sm ant-btn-icon-only ant-btn-dangerous"
            type="button"
          >
            <span
              aria-label="close"
              class="anticon anticon-close"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="close"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M563.8 512l262.5-312.9c4.4-5.2.7-13.1-6.1-13.1h-79.8c-4.7 0-9.2 2.1-12.3 5.7L511.6 449.8 295.1 191.7c-3-3.6-7.5-5.7-12.3-5.7H203c-6.8 0-10.5 7.9-6.1 13.1L459.4 512 196.9 824.9A7.95 7.95 0 00203 838h79.8c4.7 0 9.2-2.1 12.3-5.7l216.5-258.1 216.5 258.1c3 3.6 7.5 5.7 12.3 5.7h79.8c6.8 0 10.5-7.9 6.1-13.1L563.8 512z"
                />
              </svg>
            </span>
          </button>
        </div>
        <div
          class="ant-col ant-col-16"
          style="padding-left: 5px; padding-right: 5px; text-align: right;"
        >
          <div>
            1 X test item
          </div>
          <div>
            <div
              style="color: rgb(170, 170, 170); font-size: 12px;"
            >
              topping 1
            </div>
          </div>
          <div>
            <div
              style="color: rgb(170, 170, 170); font-size: 12px;"
            >
              size 1
            </div>
          </div>
        </div>
        <div
          class="ant-col ant-col-4"
          style="padding-left: 5px; padding-right: 5px; display: flex; align-items: center;"
        >
          $10
        </div>
      </div>
      <div
        class="ant-divider ant-divider-horizontal"
        role="separator"
      />
      <div
        class="ant-row ant-row-end"
        style="margin-left: -5px; margin-right: -5px;"
      >
        <div
          class="ant-col ant-col-10"
          style="padding-left: 5px; padding-right: 5px; text-align: right;"
        >
          Total:
        </div>
        <div
          class="ant-col ant-col-4"
          style="padding-left: 5px; padding-right: 5px;"
        >
          $20
        </div>
      </div>
    </div>
    <div
      style="display: flex;"
    >
      <button
        class="ant-btn ant-btn-primary ant-btn-lg"
        style="flex: 1; margin-right: 5px;"
        type="button"
      >
        <span
          aria-label="shopping-cart"
          class="anticon anticon-shopping-cart"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="shopping-cart"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="0 0 1024 1024"
            width="1em"
          >
            <path
              d="M922.9 701.9H327.4l29.9-60.9 496.8-.9c16.8 0 31.2-12 34.2-28.6l68.8-385.1c1.8-10.1-.9-20.5-7.5-28.4a34.99 34.99 0 00-26.6-12.5l-632-2.1-5.4-25.4c-3.4-16.2-18-28-34.6-28H96.5a35.3 35.3 0 100 70.6h125.9L246 312.8l58.1 281.3-74.8 122.1a34.96 34.96 0 00-3 36.8c6 11.9 18.1 19.4 31.5 19.4h62.8a102.43 102.43 0 00-20.6 61.7c0 56.6 46 102.6 102.6 102.6s102.6-46 102.6-102.6c0-22.3-7.4-44-20.6-61.7h161.1a102.43 102.43 0 00-20.6 61.7c0 56.6 46 102.6 102.6 102.6s102.6-46 102.6-102.6c0-22.3-7.4-44-20.6-61.7H923c19.4 0 35.3-15.8 35.3-35.3a35.42 35.42 0 00-35.4-35.2zM305.7 253l575.8 1.9-56.4 315.8-452.3.8L305.7 253zm96.9 612.7c-17.4 0-31.6-14.2-31.6-31.6 0-17.4 14.2-31.6 31.6-31.6s31.6 14.2 31.6 31.6a31.6 31.6 0 01-31.6 31.6zm325.1 0c-17.4 0-31.6-14.2-31.6-31.6 0-17.4 14.2-31.6 31.6-31.6s31.6 14.2 31.6 31.6a31.6 31.6 0 01-31.6 31.6z"
            />
          </svg>
        </span>
        <span>
          Checkout
        </span>
      </button>
      <button
        class="ant-btn ant-btn-primary ant-btn-lg ant-btn-icon-only ant-btn-dangerous"
        type="button"
      >
        <span
          aria-label="close"
          class="anticon anticon-close"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="close"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M563.8 512l262.5-312.9c4.4-5.2.7-13.1-6.1-13.1h-79.8c-4.7 0-9.2 2.1-12.3 5.7L511.6 449.8 295.1 191.7c-3-3.6-7.5-5.7-12.3-5.7H203c-6.8 0-10.5 7.9-6.1 13.1L459.4 512 196.9 824.9A7.95 7.95 0 00203 838h79.8c4.7 0 9.2-2.1 12.3-5.7l216.5-258.1 216.5 258.1c3 3.6 7.5 5.7 12.3 5.7h79.8c6.8 0 10.5-7.9 6.1-13.1L563.8 512z"
            />
          </svg>
        </span>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
